[stdlib] Add a new Iterable protocol with a BorrowIteratorProtocol

Azoy opened the pull request:
This is a work in progress, but should be enough to get borrowing `for in` syntax to work.

Azoy committed:
Add Iterable BorrowIteratorProtocol

Azoy committed:
Add InlineArray Iterable conformance

natecook1000 committed:
Merge branch 'main' into borrow-iterator

natecook1000 committed:
Generalize Sequence to be ~Copyable & ~Escapable

This change allows noncopyable and nonescapable Sequence types, with
room for generalizing the Element type to be ~Copyable in the future.
The new prototype adds a BorrowingIterator associated type to Sequence

This commit adds a BorrowingIteratorProtocol for generalized bulk
iteration, a BorrowingIteratorAdapter as an implementation of the new
Sequence requirement for existing sequences, and a NeverIterator for
new noncopyable/nonescapable sequence types. Conformance to Sequence
for Span and InlineArray is also included.

Issues:
- Availability checking must be off due to the availability
  requirements on the new Sequence.BorrowingIterator associated type
- Several existing sequences needed an explicit Element type alias to
  compile (note that this was during experimentation, so it may be
  possible to back out these changes)
- Some of the default implementations for generalized sequences simply
  fatal error
- InlineArray is a tricky case, due to its conditionally copyable
  design. It conforms to Sequence unconditionally, so it needs an
  unconditional implementation of (the original) Iterator associated
  type. However, when Sequence can use ~Copyable elements, that
  implementation won't be workable, since (a) it copies the inline
  array, and (b) it implements the `next() -> Element?` method, which
  can't return a noncopyable element.

airspeedswift commented:
@swift-ci build toolchain

airspeedswift committed:
Mark borrowing iterator types frozen and methods transparent

airspeedswift commented:
with https://github.com/swiftlang/swift-experimental-string-processing/pull/843
@swift-ci build toolchain

airspeedswift committed:
Update availability for Deque's typealias

airspeedswift committed:
More typealias availability

slavapestov commented on stdlib/private/StdlibCollectionUnittest/MinimalCollections.swift:
```diff
@@ -83,6 +83,8 @@ public enum UnderestimatedCountBehavior {
 ///
 /// This sequence is consumed when its generator is advanced.
 public struct MinimalSequence<T> : Sequence, CustomDebugStringConvertible {
+  public typealias Element = T
```
Here it might be worth just renaming the generic parameter T to Element, for readability in any case

natecook1000 committed:
Make elements equal impl slightly more realistic

natecook1000 committed:
Merge branch 'main' into borrow-iterator

natecook1000 committed:
Work on ~Copyable elements (not building yet)

natecook1000 committed:
Fix the BorrowingIteratorAdapter type

Because the BorrowingIteratorAdapter type still required Copyable
elements, it was failing to be allowed as the default BorrowIterator
type for any sequences. However, just marking the `Iterator.Element`
type as `~Copyable` is impossible, since IteratorProtocol has a
copyable element type, and switching to being generic over a sequence
with a noncopyable element type fails with this error:

Cannot suppress '~Copyable' on generic parameter 'S.Element' defined in outer scope

This change resolves things by adding `Element: ~Copyable` as a
second generic parameter and equating the new type and
`Iterator.Element`.

natecook1000 committed:
Back out new type aliases for existing sequences
